### ISSUE TITLE
PureBasic Help: Enforce ISO-8859-1/Latin-1 in Sources Encoding

### ISSUE THREAD

**Issue by Naheulf** [fantaisie-software/purebasic#152]
In #135 I discover that DocMaker does not support utf8. After that @tajmone Suggested Immediate Repository Actions. However they are the exact opposite of what the issue was create for. So I create a new one while waiting for DocMaker to accept utf8.

@tajmone I let you copy theses action in thi issue. So you can link them with the [PureBasic Documentation](https://github.com/fantaisie-software/purebasic/projects/2) project.

**Comment by tajmone** [comment]
Enforcing ISO-8859-1 on the repository is not possible, Git does not support this.

Since EClint doesn't handle properly Latin1 validation, what we can do is tweak the `validate.sh` script and validate the encoding of all `.txt` files inside the `Documentation/` directory tree using **iconv** (i.e. convert each file from and to ISO-8859-1, dump the output to `nul` and check for exit code errors, and report the corrupted file when the Travis CI build fails).

This is going to slow down considerably the build validation, but it shouldn't be an issue really.

What worries me most is the end-users side. We can set in `.editorconfig` an ISO-9959-1 encoding rule for the DocMaker `.txt` files, but we won't have any guarantee that even editors supporting EditorConfig will abide by them (features support can vary among editors plug-ins) — and of course, end users working with an editor/IDE that doesn't support EditorConfig will have to manually handle the encoding against their editor's default settings.

Preventing corrupted files from being merged into the repository is only one side of the problem; the other one is ensuring it doesn't become too frustrating for contributors to edit those files.

We should probably find a good free and open source editor that can handle well ISO-8859-1/Latin1, and instruct contributors to use that editor, and how to correctly configure it. I'm thinking of Notepad++, which if I remember correctly is quite good at handling the different encodings, since its roots go back to the CodePages and ISO era. But I need to check if it preserves the encoding with paste operations, etc.

We might also need to create some custom settings file to offer for download — but my main concern is that we would be affecting how the `.txt` extension is handled by the editor (which is the most common plain-text extension used). Unless Notepad++ supports projects settings, in which case we could simply include a project configuration in the repository, and the editor will pick it up.

**Comment by Naheulf** [comment]
When Notepad++ open a file it try to detect the file encoding and keep it in memory. At save time it convert the working text to the memorized encoding (and use a substitution char if the character is not available in the target charset).
However this behaviour will be useless if the file contain only pure ASCII chars since Notepad++ will detect it as UTF8...

Another way to handle it is to use a control line to check if encoding has changed.

For example we can include:
```purebasic
;
;  DocMaker source files MUST be encoded in ISO-8859-1/Latin-1/Windows-1252
;  This whole comment is used to detect and enforce file encoding.
;  ** Do NOT remove or edit them. **
;  Ref Line: e a i o E U (simple letters without diacritics)
;  Test Line: é à î ö Ê Ü (same letters with one diacritic)
;
```
in all DocMaker input files. And we can get all files without the test line encoded it the good format with:
```shell
#!/bin/bash

basedir=$(dirname $0)
tmpLog=$(mktemp)

# **************************************************
# A. Check DocMaker input files encoding
# **************************************************
# Currently DocMaker only work with files encoded in
# ISO-8859-1/Latin-1/Windows-1252.
#
# To ensure that we need to check the presence of a
# test line including specials characters encoded in
# this charset.

echo -e "\n\033[34;1m================================================"
echo -e "\033[33;1mCheck for DocMaker input files encoding"
echo -e "\033[34;1m================================================\033[0m"
IFS_COPY=$IFS
IFS=$(echo -en "\n\b")
LC_CTYPE_COPY=$LC_CTYPE
LC_CTYPE=C

grep -l '^ *@' $(grep -rL  $';  Test Line: \xE9 \xE0 \xEE \xF6 \xCA \xDC (same letters with one diacritic)' --include="*.txt" ${basedir}) >> $tmpLog
if [ $? -eq 0 ] ; then
	echo -e "\033[31;1m~~~ ERROR! ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
	echo -e "\033[31;1mThe following files have encoding issue:\n\033[33;1m"
	cat $tmpLog
	rm $tmpLog
	echo -e "\033[31;1m\nPlease clean up your sources and fix the problem."
	echo -e "\033[31;1m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
	echo -e "\033[31;1m/// Aborting All Tests ///\033[0m"
	exit 1
fi

LC_CTYPE=$LC_CTYPE_COPY # Restore original LC_CTYPE!
IFS=$IFS_COPY # Restore original IFS!

echo -e "\033[32;1m/// Test Passed ///\033[0m"
rm $tmpLog
```
Note the grep finding '^ *@' to ensure the file has command for DocMaker.

**Comment by tajmone** [comment]
> However this behaviour will be useless if the file contain only pure ASCII chars since Notepad++ will detect it as UTF8.

Indeed, that's how the encoding problem went unnoticed when I edited the English documentation (goo job I didn't add any special chars, or it would have broken the source). The same problem should occur with most editors, since UTF-8 is the assumed encoding today.

> Another way to handle it is to use a control line to check if encoding has changed.

I like this idea, the control line itself should suffice as a means for most editors/IDEs to auto-detect the ISO encoding, and your validation script should be much faster that checking the whole file via `inconv`, since it can test a single specific line with each file — although only a full `inconv` processing can guarantee the whole file integrity.

```language
;  DocMaker source files MUST be encoded in ISO-8859-1/Latin-1/Windows-1252
```

ISO-8859-1 and Latin-1 are the same encoding, but Windows-1252 is not. This confusion about Windows-1252 being equivalent to ISO-8859-1 has a long history and has been the cause of many text corruption issues. Although the two encodings overlap for the most part, they also differ in some entries:

- [Comparing Characters in Windows-1252, ISO-8859-1, ISO-8859-15]
- [Encoding Problem: Treating UTF-8 Bytes as Windows-1252 or ISO-8859-1]
- [UTF-8 Encoding Debugging Chart]





[Comparing Characters in Windows-1252, ISO-8859-1, ISO-8859-15]: https://i18nqa.com/debug/table-iso8859-1-vs-windows-1252.html

[Encoding Problem: Treating UTF-8 Bytes as Windows-1252 or ISO-8859-1]: https://i18nqa.com/debug/bug-utf-8-latin1.html
[UTF-8 Encoding Debugging Chart]: https://i18nqa.com/debug/utf8-debug.html

**Comment by Naheulf** [comment]
> ```
> ;  DocMaker source files MUST be encoded in ISO-8859-1/Latin-1/Windows-1252
> ```
> 
> ISO-8859-1 and Latin-1 are the same encoding, but Windows-1252 is not. This confusion about Windows-1252 being equivalent to ISO-8859-1 has a long history and has been the cause of many text corruption issues. Although the two encodings overlap for the most part, they also differ in some entries:

That right. So I did even more research. And now I can confirm the Encoding used in the doc: it's Windows-1252. Let me explain why this one and not another.

In all three `String.txt` files there an example (I replaced the character in the string by the replacement character):
```
  Debug Asc("�") ; will print '8364' = 20AC in hexadecimal
```
In Unicode world the character with value "8364" is "€". Since the file the "�" character is coded with $80 in the file, that mean : the charset map the $80 value to the Euro sign. And the only on charset who do that is... Windows-1252.

To confirm it I use grep with `$'[\x7F-\x9F\xA4\xA6\xA8\xB4\xB8\xBC\xBD\xBE]'` to find all characters variable between theses encodings. And for all tested results the character make sense only in Windows-1252.

I fund ONE exception: `SpiderBasic\English\Reference\license_application.txt`. This file is encoded in UTF-8 because it use the `ń` char only available in Unicode. And, as expected, it show wrong in the output file https://www.spiderbasic.com/documentation/reference/license_application.html


*****

This works pretty well on Windows(8.1 x64), but since the charset came from Microsoft... it doesn't works on Linux (Lubuntu x64). And we can show strange characters, even in the english files (most for quotes and hyphens).
By the way the first example above doesn't works since Linux read the $80 character as  "" (or "" in unicode)

So, if we want to make the help working on all OS we also need to remove all Windows-1252 specific chars or (ab)use of `@OS` directives to match all OS charset. (Except for the example above I think it's a bad idea)

**Comment by tajmone** [comment]
> That right. So I did even more research. And now I can confirm the Encoding used in the doc: it's Windows-1252.

Wow, impressive research work @Naheulf, indeed.

> In all three String.txt files there an example (I replaced the character in the string by the replacement character):

I believe you're referring to the [`Documentation/English/String.txt`][String.txt] source (and its corresponding file in German and French too):

```purebasic
  unicode$="€€ "
  Debug Asc(unicode$) ; will print '8364'
```

this specific example is found in the documentation of the [Asc()] command.

Yes, if I open the file in Notepad++ and play around with the encodings I can see that only Windows-1252 shows the &eur; symbols correctly (ISO-8859-1 produced a <kbd>PAD</kbd> substitution char).

But then, when I build the HTML documentation locally I notice that the example is wrongly encoded in the final document (built under `Documentation/English/HTML/string/asc.html`):

```purebasic
  unicode$="АА "
  Debug Asc(unicode$) ; will print '8364'
```

(**NOTE** — in [the online documentation on PB website][Asc()], that  page is correctly encoded)

So my guess is that although the source file is currently encoded as Windows-1252, it should have been in ISO-8859-1, and that it might have got corrupted somewhere along the line when the repository was moved from SVN to Git.

This is not unusual, and I've encountered similar problems with the ISO-8859-1 repositories I've worked with, where many files got corrupted during their migration to Git/GitHub, and I had to fix them manually.

I still think that DocMaker expects ISO-8859-1 sources, but we still need a confirmation by @fantaisie-software on this.

Another problem is that the HTML pages don't contain any encoding directives in their source:

```html
<html><head><title>Asc</title></head>
<body bgcolor="#FFFFDF" link="#009999" vlink="#006666" alink="#006666">
```
... so how these characters are ultimately represented might largely depend on the browser settings, and for CHM files on the default system language. Without DocMaker adding a proper encoding specification in the generated HTML output, chances are that end users might see different results depending on their OS settings.


> So, if we want to make the help working on all OS we also need to remove all Windows-1252 specific chars

I definitely agree on this. If shifting to UTF-8 is not a viable option (i.e. due to having to update DocMaker itself), and we therefore need to stick with a single-character encoding, then Latin-1/ISO-8859-1 seems the most reasonable choice since PB Help is only available in EN, DE and FR locales.




[String.txt]: https://github.com/fantaisie-software/purebasic/blob/001b3f3/Documentation/English/String.txt#L48 "View source file"

[Asc()]: https://www.purebasic.com/documentation/string/asc.html "View at PB online documentation"

**Comment by Naheulf** [comment]
> > In all three String.txt files there an example (I replaced the character in the string by the replacement character):
> 
> I believe you're referring to the [`Documentation/English/String.txt`](https://github.com/fantaisie-software/purebasic/blob/001b3f3/Documentation/English/String.txt#L48) source (and its corresponding file in German and French too):

Right. There only three `String.txt` in the repo anyway. On for each language. However my example came from [`line 46`](https://github.com/fantaisie-software/purebasic/blob/001b3f3/Documentation/English/String.txt#L46), not 48. 



> But then, when I build the HTML documentation locally I notice that the example is wrongly encoded in the final document (built under `Documentation/English/HTML/string/asc.html`):
> 
> ```
>   unicode$="АА "
>   Debug Asc(unicode$) ; will print '8364'
> ```
> 
> (**NOTE** — in [the online documentation on PB website](https://www.purebasic.com/documentation/string/asc.html), that page is correctly encoded)

For me both build and online doc display `"AA "` in Firefox and Chrome. And both build and online doc display `"€€ "` in IE and Notepad++.
![Headbang](https://media.tenor.com/images/4b850ab454938ae62733c78d8b0edc61/tenor.gif)

**Comment by tajmone** [comment]
I will keep repeating it: working with single-character encodings today is a real nightmare!

Whenever you think that you've solved one problem, there's another one awaiting around the corner!

It's just to hard to ensure that all the tools work as expected, and the situation becomes even more tiring in projects where you have contributors using different OSs and tools.

**Comment by tajmone** [comment]
> For me both build and online doc display "AA " in Firefox and Chrome. And both build and online doc display "€€ " in IE and Notepad++.

That's because:

1. Firefox and Chrome are modern browsers, and they assume UTF-8 as the default encoding (they don't even offer an option to switch to the legacy encodings, not Chrome at least).
2. IE is a legacy browser, and (being from MS) it defaults to Windows-1252 (not ISO-8859-1). The same applies to CHM files, which are viewed via the WebBrowser component, but in IE7 emulation mode.
3. Notepad++ is particularly good at detecting legacy encodings, because its codebase dates back to the era in which these were still used. Modern editors don't invest so much energy on encoding detection.

**Comment by tajmone** [comment]
I'm also converting this Issue into a Discussion, because there are too many open questions to consider it a clear directive on what to do with the repository. Once all aspects of the encoding problem are considered, and we have a clean-cut agreed upon course of action, we'll create a new and concise Issue with the required actions to fix the problem.
